compiler/src/iree/compiler/Codegen/Common/GPU/GPUConvertToCoalescedDMA.cpp

```diff
@@ -0,0 +1,168 @@
+
+    SmallVector<OpFoldResult> tileSizes;
+    SmallVector<OpFoldResult> lowerBounds;
+    SmallVector<OpFoldResult> upperBounds;
+    SmallVector<OpFoldResult> steps;
+
+    Location loc = gatherOp.getLoc();
+    for (int64_t i = 0; i < rank; ++i) {
+      OpFoldResult tileSize = subgroupTileSizes[i];
+      tileSizes.push_back(tileSize);
+      lowerBounds.push_back(rewriter.getIndexAttr(0));
+      upperBounds.push_back(rewriter.getIndexAttr(initType.getDimSize(i)));
+      steps.push_back(tileSize);
+    }
+
+    SmallVector<Attribute> mapping;
+    for (int64_t i = 0; i < rank; ++i) {
+      mapping.push_back(gpu::GPUWarpMappingAttr::get(
+          rewriter.getContext(),
+          static_cast<gpu::MappingId>(
+              static_cast<int>(gpu::MappingId::LinearDim0) + (rank - 1 - i))));
+    }
+
+    auto forallOp =
+        rewriter.create<scf::ForallOp>(loc, lowerBounds, upperBounds, steps,
+                                       init, rewriter.getArrayAttr(mapping));
+
+    Block *body = forallOp.getBody();
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(body);
+
+    SmallVector<Value> ivs(body->getArguments().begin(),
+                           body->getArguments().begin() + rank);
+    Value sharedOut = body->getArguments()[rank];
+
+    SmallVector<OpFoldResult> offsets;
+    for (Value iv : ivs) {
+      offsets.push_back(iv);
+    }
+    SmallVector<OpFoldResult> strides(rank, rewriter.getIndexAttr(1));
+
+    auto extractedIndices = rewriter.create<tensor::ExtractSliceOp>(
+        loc, indices, offsets, tileSizes, strides);
+
+    auto indicesType = cast<RankedTensorType>(extractedIndices.getType());
+    auto indexIndicesType =
+        RankedTensorType::get(indicesType.getShape(), rewriter.getIndexType());
+    Value indexIndices = rewriter.create<arith::IndexCastOp>(
+        loc, indexIndicesType, extractedIndices.getResult());
+
+    auto inParallelOp = cast<scf::InParallelOp>(body->getTerminator());
+
+    Block &inParallelBlock = inParallelOp.getRegion().front();
+    rewriter.setInsertionPointToStart(&inParallelBlock);
+
+    rewriter.create<IREE::GPU::CoalescedGatherDMAOp>(
+        loc, sharedOut.getType(), indexIndices, source, sharedOut);
+
+    rewriter.replaceOp(gatherOp, forallOp.getResults());
+
+    return success();
+  }
+};
+
+struct GPUConvertToCoalescedDMAPass final
+    : impl::GPUConvertToCoalescedDMAPassBase<GPUConvertToCoalescedDMAPass> {
+  using GPUConvertToCoalescedDMAPassBase::GPUConvertToCoalescedDMAPassBase;
+  void runOnOperation() override {
+    FunctionOpInterface funcOp = getOperation();
+
+    LLVM_DEBUG(llvm::dbgs()
+               << "Running GPUConvertToCoalescedDMAPass on function: "
+               << funcOp.getName() << "\n");
+
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+
+    patterns.add<ConverGatherOpToCoalescedDMA>(context);
+
+    if (failed(applyPatternsGreedily(funcOp, std::move(patterns)))) {
```

Comment:
Use `walkAndApplyPatterns` instead since it's a single pattern that only applies a single time per op.
```suggestion
    if (failed(walkAndApplyPatterns(funcOp, std::move(patterns)))) {
```